libgalois/src/GraphTopology.cpp

```diff
@@ -435,6 +435,182 @@ katana::EdgeTypeAwareTopology::MakeFrom(
       pg, edge_type_index, e_topo, std::move(per_type_adj_indices)});
 }
 
+std::unique_ptr<katana::ProjectedShuffleTopology>
+katana::ProjectedShuffleTopology::MakeProjectedGraph(
+    const katana::PropertyGraph* pg,
+    const std::vector<std::string>& node_properties,
+    const std::vector<std::string>& edge_properties) {
+  KATANA_LOG_DEBUG_ASSERT(pg);
+
+  const auto& topology = pg->topology();
+  if (topology.empty()) {
+    ProjectedShuffleTopology pst;
+    return std::make_unique<ProjectedShuffleTopology>(std::move(pst));
+  }
+
+  // calculate number of new nodes
+  uint32_t num_new_nodes = 0;
+  uint32_t num_new_edges = 0;
+
+  katana::DynamicBitset bitset_nodes;
+  bitset_nodes.resize(topology.num_nodes());
+
+  if (node_properties.empty()) {
+    num_new_nodes = topology.num_nodes();
+    // set all nodes
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      bitset_nodes.set(src);
+    });
+  } else {
+    std::set<katana::EntityTypeID> node_entity_type_ids;
+
+    for (auto node_prop : node_properties) {
+      auto entity_type_id = pg->GetNodeEntityTypeID(node_prop);
+      node_entity_type_ids.insert(entity_type_id);
+    }
+
+    katana::GAccumulator<uint32_t> accum_num_new_nodes;
+    katana::GAccumulator<uint64_t> accum_num_new_edges;
+
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      for (auto type : node_entity_type_ids) {
+        if (pg->DoesNodeHaveType(src, type)) {
+          accum_num_new_nodes += 1;
+          accum_num_new_edges += std::distance(
+              topology.edges(src).begin(), topology.edges(src).end());
+          bitset_nodes.set(src);
+          return;
+        }
+      }
+    });
+    num_new_nodes = accum_num_new_nodes.reduce();
+    num_new_edges = accum_num_new_edges.reduce();
+  }
+
+  // fill old to new nodes mapping
+  NUMAArray<Node> old_to_new_nodes_mapping;
+  NUMAArray<Node> new_to_old_nodes_mapping;
+
+  old_to_new_nodes_mapping.allocateInterleaved(topology.num_nodes());
+  new_to_old_nodes_mapping.allocateInterleaved(num_new_nodes);
+
+  Node node_idx{0};
+  for (Node src = 0; src < topology.num_nodes(); src++) {
```

Comment:
We're gonna need a parallel implementation of this loop. See the implementation of `MakeNodeSortedTopo` to see how you can populate these mappings in parallel. 
Also, perhaps we won't need old to new mapping, but if you want it, see `MakeNodeSortedTopo`.